Support deletion of CloudObjectStoreContainer

miha-plesko opened the pull request:
CloudObjectStoreContainer now supports async task processing
therefore we support the first operation with this commit: destroy.
See video attached to the manage-ui-classic PR (link below).

Links:
* https://github.com/ManageIQ/manageiq/pull/13965 (already merged)
* https://github.com/ManageIQ/manageiq-ui-classic/pull/420 (GUI for this operation)

@miq-bot add_label enhancement
@miq-bot assign @blomquisg

miha-plesko commented:
@blomquisg  @Ladas could you take a look at this, please? If possible, I'd like to have this first operation merged soon so that I can then add more operation to the CloudObjectStoreContainer :D

Ladas commented on app/models/manageiq/providers/amazon/inventory/parser/storage_manager/s3.rb:
```diff
@@ -28,6 +28,7 @@ def process_containers
   def parse_container(bucket)
     uid = bucket['name']
     {
+      :type                  => self.class.container_type,
```
seems like there is no type in the DB, we will need a migration to add it

Ladas replied:
ah, you are already adding it, so :+1:

durandom replied:
why cant this be just `self.class.name` ?

miha-plesko replied:
I think  `self.class.name` would evaluate in
```ruby
ManageIQ::Providers::Amazon::Inventory::Parser::StorageManager::S3
```
while I need
```ruby
ManageIQ::Providers::Amazon::Inventory::Parser::StorageManager::S3::CloudObjectStoreContainer
```

Ladas commented:
@miha-plesko looks great, only thing I am missing is a 'supports' for the destroy, it should be also in the UI PR, @durandom can you point @miha-plesko ?

durandom commented on .gitignore:
```diff
@@ -14,3 +14,5 @@
 
 /spec/manageiq
+
+.idea/*
```
you can add that to your global .gitignore - thats what I do :)

miha-plesko commented:
@Ladas, @durandom, gentlemen, thanks for pointing me to 'supports' funcionality. I've added it both to menu definition (details page) and to controller (details page + list page). Also I've aligned flash message texts with current ones in `.po` and added tests for them.

Looking forward for ya opinion!

durandom commented on app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_container.rb:
```diff
@@ -0,0 +1,10 @@
+class ManageIQ::Providers::Amazon::StorageManager::S3::CloudObjectStoreContainer < ::CloudObjectStoreContainer
+  include_concern 'Operations'
+
+  supports :delete
```
can a `CloudObjectStoreContainer` always be deleted? Or are there some prerequisites?

miha-plesko replied:
As discussed on gitter, lets let it fail if e.g. there is no permission. We will fine-tune it later.

durandom commented on app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_container/operations.rb:
```diff
@@ -0,0 +1,9 @@
+module ManageIQ::Providers::Amazon::CloudManager::CloudObjectStoreContainer::Operations
+  extend ActiveSupport::Concern
+
+  def raw_destroy
```
`destroy` vs `delete` - it seems `raw_delete` is used more often than `raw_destroy`
`delete` would have the benefit of aligning more with `supports_delete`

Ladas commented on app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_container/operations.rb:
```diff
@@ -0,0 +1,9 @@
+  def raw_destroy
+    raise "CloudObjectStoreContainer has no #{ui_lookup(:table => "ext_management_systems")}, unable to destroy it" unless ext_management_system
+    with_provider_object(&:delete!)
+    destroy
```
we should probaly leave MIQ DB destroy to refresh, this could optionaly set a state do 'deleting', but there is not state on S3 object, right?

miha-plesko commented:
@durandom @Ladas thanks for the constructive comments. I've fixed the issues and repushed.

sergio-ocon commented:
@miha-plesko @durandom, there is something missing in this PR? We have some PR depending on this one, we need ti merged quick

Ladas commented:
looks good now, but we need ManageIQ/manageiq#13965 merged first, so the CI passes

sergio-ocon commented:
Yes, I've seen that that PR is still waiting to be merged, but I don't see anything blocking it

gberginc commented:
@miha-plesko please close and reopen to check if this PR passes tests now that https://github.com/ManageIQ/manageiq/pull/13965 was merged.

miha-plesko commented:
@durandom after merging dependent PR from core miq, tests here turn green. Could we push it forward now?

sergio-ocon commented:
Hi @durandom , what do you think?

durandom commented on app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_container.rb:
```diff
@@ -0,0 +1,17 @@
+    unless ext_management_system
+      unsupported_reason_add(
+        :delete,
+        "CloudObjectStoreContainer has no #{ui_lookup(:table => "ext_management_systems")}, unable to delete it"
```
i18n should be like

```ruby
    unsupported_reason_add(:delete_security_group, _("The Security Group is not connected to an active %{table}") % {
        :table => ui_lookup(:table => "ext_management_systems")
      })
```

miha-plesko replied:
👍

durandom commented on spec/models/manageiq/providers/amazon/storage_manager/s3/stubbed_refresher_spec.rb:
```diff
@@ -4,14 +4,14 @@
-    end
+  before do
+    _guid, _server, zone = EvmSpecHelper.create_guid_miq_server_zone
+    @ems                 = FactoryGirl.create(:ems_amazon, :zone => zone)
```
for future code, always prefer `let` over instance vars in before blocks. The latter will lead to object instantiated even if they are not used in the specs.

miha-plesko replied:
Great idea, thanks. I've implemented it here already.

durandom commented on spec/factories/cloud_object_store.rb:
```diff
@@ -0,0 +1,17 @@
+    bucket.sequence(:name) { |n| "stubbed-name-#{n}" }
+    bucket.sequence(:ems_ref) { |n| "stubbed-ems-ref-#{n}" }
+
+    after(:create) do |b|
```
bucket was too long? 😛

miha-plesko replied:
Rubocop pointed his gun to me saying "Shadowing outer local variable - bucket". But I too prefer explicit names so fuck rubocop, I'm using `bucket` here. Thanks for pointing this out.

durandom commented on app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_container/operations.rb:
```diff
@@ -0,0 +1,7 @@
+module ManageIQ::Providers::Amazon::CloudManager::CloudObjectStoreContainer::Operations
```
Yeah, because of such small files I'm not a big fan of spreading the code across modules.

So, if you like, you can inline this here in the aws gem 😄  

https://blog.codeship.com/good-module-bad-module/

miha-plesko replied:
I do like 😄

durandom commented:
missing service model https://github.com/ManageIQ/manageiq/pull/14164

sergio-ocon commented:
:clap:
